### PR TITLE
feat(project): add migration endpoint (DSP-1666)

### PR DESCRIPTION
**Pull request by mdelez**
resolves DSP-1666

### CODE REVIEW

**Comment by mdelez** [comment]
@BalduinLandolt this PR is a setup for https://dasch.myjetbrains.com/youtrack/issue/DSP-1614. I've added a new endpoint specifically for the migration of current projects to the new system. Currently it's an exact copy of the `CreateProject` method. The reason being that the `CreateProject` method will change in the aforementioned youtrack issue so that the short code cannot be specified and will instead be auto-generated when the project is created. The migration endpoint will allow you to specify the short code and will be behind quite restrictive permissions once those are set up.

File: services/admin/backend/api/handler/project.go [hunk 1]
```diff
@@ -377,6 +377,110 @@ func deleteProject(service project.UseCase) http.Handler {
...
 }
 
+func migrateProject(service project.UseCase) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		errorMessage := "Error creating project"
+		var input struct {
+			ShortCode   string `json:"shortCode"`
+			ShortName   string `json:"shortName"`
+			LongName    string `json:"longName"`
+			Description string `json:"description"`
```
Comment (mpro7) [inline services/admin/backend/api/handler/project.go:387]: Since this struct is used over few functions here it should be defined once and reused.

**Review by mpro7 — COMMENTED** [review]
Isn't there better way to combine the migration function with `createProject`?

**Review by BalduinLandolt — APPROVED** [review]
The code, being a copy of the `createProject` endpoint, can easily be approved, I guess.

However I'm not 100% sure the solution of separating those endpoints will prove practical: My gut tells me that we'll end up using the migration endpoint all the time (maybe not in the long term, but middle term I guess) because it might prove hindering with our workflows of accepting projects. I imagine CS will want to assign projects with short codes before they have all the information necessary to actually create the project. Or it may happen that projects need a short code when applying for funding, but we don't want to create the project before we are sure they receive funding.  
I think we have to further discuss this at some point, or at least keep it in the back of our minds that this may be an issue.

**Comment by mdelez** [comment]
> The code, being a copy of the `createProject` endpoint, can easily be approved, I guess.
> 
> However I'm not 100% sure the solution of separating those endpoints will prove practical: My gut tells me that we'll end up using the migration endpoint all the time (maybe not in the long term, but middle term I guess) because it might prove hindering with our workflows of accepting projects. I imagine CS will want to assign projects with short codes before they have all the information necessary to actually create the project. Or it may happen that projects need a short code when applying for funding, but we don't want to create the project before we are sure they receive funding.
> I think we have to further discuss this at some point, or at least keep it in the back of our minds that this may be an issue.

My first thought was to simply make the short code optional but Go doesn't support that so then I was thinking of a hacky solution with an additional boolean parameter that would control whether or not the provided short code should be used or it should be auto-generated. I'm also not a fan of temporary code because if often turns into permanent code. For now I'll stick with the two different methods for migration and creation but we should discuss it either this week or next week. Otherwise, I fear it will be forgotten.

**Comment by BalduinLandolt** [comment]
> 
> My first thought was to simply make the short code optional but Go doesn't support that so then I was thinking of a hacky solution with an additional boolean parameter that would control whether or not the provided short code should be used or it should be auto-generated. I'm also not a fan of temporary code because if often turns into permanent code. For now I'll stick with the two different methods for migration and creation but we should discuss it either this week or next week. Otherwise, I fear it will be forgotten.

Too bad Go doesn't support optional params...  
Would you consider it hacky too, to 1) check if a shortcode was passed via API, 2) if so, use that, and if not, use the string null-value, and pass this "pseudo-optional" shortcode, and 3) when creating the project, check if shortcode is null-value, and if so, auto-generate one.  
(That's another question: how can we reasonably auto-create shortcodes?)

**Pull request merged**